pkg/query-service/agentConf/manager.go

```diff
@@ -0,0 +1,217 @@
+package agentConf
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/jmoiron/sqlx"
+	"go.signoz.io/signoz/pkg/query-service/app/opamp"
+	filterprocessor "go.signoz.io/signoz/pkg/query-service/app/opamp/otelconfig/filterprocessor"
+	tsp "go.signoz.io/signoz/pkg/query-service/app/opamp/otelconfig/tailsampler"
+	"go.uber.org/zap"
+	yaml "gopkg.in/yaml.v3"
+)
+
+var m *Manager
+
+func init() {
+	m = &Manager{}
+}
+
+type Manager struct {
+	Repo
+	// lock to make sure only one update is sent to remote agents at a time
+	lock uint32
+}
+
+// Ready indicates if Manager can accept new config update requests
+func (mgr *Manager) Ready() bool {
+	if mgr.lock != 0 {
+		return false
+	}
+	return opamp.Ready()
+}
+
+func Initiate(db *sqlx.DB, engine string) error {
+	m.Repo = Repo{db}
+	return m.initDB(engine)
+}
+
+// Ready indicates if Manager can accept new config update requests
+func Ready() bool {
+	return m.Ready()
+}
+
+func GetLatestVersion(ctx context.Context, elementType ElementTypeDef) (*ConfigVersion, error) {
+	return m.GetLatestVersion(ctx, elementType)
+}
+
+func GetConfigVersion(ctx context.Context, elementType ElementTypeDef, version int) (*ConfigVersion, error) {
+	return m.GetConfigVersion(ctx, elementType, version)
+}
+
+func GetConfigHistory(ctx context.Context, typ ElementTypeDef) ([]ConfigVersion, error) {
+	return m.GetConfigHistory(ctx, typ)
+}
+
+// StartNewVersion launches a new config version for given set of elements
+func StartNewVersion(ctx context.Context, eleType ElementTypeDef, elementIds []string) (*ConfigVersion, error) {
+
+	if !m.Ready() {
+		// agent is already being updated, ask caller to wait and re-try after sometime
+		return nil, fmt.Errorf("agent updater is busy")
+	}
```

Comment:
If the suggestion for caller is to wait and retry the I would suggest make channel and block here instead returning error.

Reply:
The function would return ERROR to the caller.